### DISCUSSION TITLE
[Deployment] Error - The code hash is not known

### DISCUSSION THREAD

**Discussion by 4NNNN** [zkSync-Community-Hub/zksync-developers#125] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

latest

### zksync-web3 Version

0.14.3

### Hardhat.config.ts

import { HardhatUserConfig } from "hardhat/config";

import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";
import "@matterlabs/hardhat-zksync-verify";

const config: HardhatUserConfig = {
  zksolc: {
    version: "1.3.10", // Use latest available in https://github.com/matter-labs/zksolc-bin/
    compilerSource: "binary",
    settings: {
      isSystem: true,
    },
  },
  defaultNetwork: "zkSyncTestnet",
  networks: {
    hardhat: {
      zksync: true,
    },
    zkSyncTestnet: {
      url: "https://zksync2-testnet.zksync.dev",
      ethNetwork: "https://goerli.infura.io/v3/0100a2dba004414e9b71717ef18fe197" , // e.g. alchemy url
      //url: "http://localhost:3050",
      //ethNetwork: "http://localhost:8545",
      zksync: true,
      verifyURL: 'https://zksync2-testnet-explorer.zksync.dev/contract_verification'
    },
  },
  solidity: {
    version: "0.8.16",
  },
};

export default config;

### Deployment Script (WITHOUT PRIVATE KEY)

import { utils, Wallet, Provider, EIP712Signer, types } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";

export default async function (hre: HardhatRuntimeEnvironment) {
  const provider = new Provider("https://testnet.era.zksync.dev");
  const wallet = new Wallet("").connect(provider);
  const factoryArtifact = await hre.artifacts.readArtifact("GuardianStorage");
  const accountArtifact = await hre.artifacts.readArtifact("ContractDeployer");
  const accountContract = "0x0000000000000000000000000000000000008006"
  const guard = factoryArtifact.bytecode;
  //console.log(bytecode);
  const account = new ethers.Contract(
    accountContract,
    accountArtifact.abi,
    wallet
  );

  function generateBytecodeHash(bytecode) {
    // Calculate the sha256 hash of the bytecode
    const sha256Hash = crypto.createHash('sha256').update(bytecode).digest();

    // Get the last 28 bytes of the sha256 hash
    const last28Bytes = sha256Hash.slice(-28);

    // Get the length of the bytecode in 32-byte words
    const bytecodeLength = Math.ceil(bytecode.length / 32);

    // Create a buffer for the bytecode hash
    const bytecodeHash = Buffer.alloc(32);

    // Set the first 2 bytes to the version of the bytecode hash format
    bytecodeHash[0] = 1;
    bytecodeHash[1] = 0;

    // Set the second 2 bytes to the length of the bytecode
    bytecodeHash.writeUInt16BE(bytecodeLength, 2);

    // Set the rest of the bytes to the last 28 bytes of the sha256 hash
    last28Bytes.copy(bytecodeHash, 4);

    return bytecodeHash;
  }
 console.log(generateBytecodeHash(guard).toString('hex'));
 const ash = "0x"+ generateBytecodeHash(guard).toString('hex');

 let aatx = await account.create("0x0000000000000000000000000000000000000000000000000000000000000000",ash,"0x");
  console.log(aatx);
}

### Contract Code

// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.8.12 <0.9.0;

import "./IGuardianstorage.sol";
import "contracts/TwoUserMultisig.sol";

/**
 * @title GuardianStorage
 * @notice Contract storing the state of wallets related to guardians.
 * This contract only contains basic setters/getters, only enabled modules for a wallet can modify its state
 * @author CANDIDE Labs team
 */
contract GuardianStorage is IGuardianStorage {

    address internal constant SENTINEL_GUARDIANS = address(0x1);

    struct GuardianStorageEntry {
        // the list of guardians
        mapping(address => address) guardians;
        // guardians count
        uint256 count;
        // recovery threshold
        uint256 threshold;
    }

    mapping (address => GuardianStorageEntry) internal entries;

 /**
     * @dev Throws if the caller is not an enabled module.
     * @param _wallet The target wallet.
     */
    modifier onlyModule(address _wallet) {
        // solhint-disable-next-line reason-string
        require(TwoUserMultisig(payable(_wallet)).isModuleEnabled(msg.sender), "GS: methold only callable by an enabled module");
        _;
    }

    /**
     * @dev Lets an authorised module add a guardian to a wallet.
     * @param _wallet The target wallet.
     * @param _guardian The guardian to add.
     */
    function addGuardian(address _wallet, address _guardian) external onlyModule(_wallet) {
        require(_guardian != address(0) && _guardian != SENTINEL_GUARDIANS && _guardian != _wallet, "GS: invalid guardian");
        require(!TwoUserMultisig(payable(_wallet)).isOwner(_guardian), "GS: guardian cannot be an owner");
        GuardianStorageEntry storage entry = entries[_wallet];
        require(entry.guardians[_guardian] == address(0), "GS: duplicate guardian");
        if (entry.count == 0){
            entry.guardians[SENTINEL_GUARDIANS] = _guardian;
            entry.guardians[_guardian] = SENTINEL_GUARDIANS;
        }else{
            entry.guardians[_guardian] = entry.guardians[SENTINEL_GUARDIANS];
            entry.guardians[SENTINEL_GUARDIANS] = _guardian;
        }
        entry.count++;
    }

    /**
     * @dev Lets an authorised module revoke a guardian from a wallet.
     * @param _wallet The target wallet.
     * @param _prevGuardian Guardian that pointed to the guardian to be removed in the linked list
     * @param _guardian The guardian to revoke.
     */
    function revokeGuardian(address _wallet, address _prevGuardian, address _guardian) external onlyModule(_wallet) {
        GuardianStorageEntry storage entry = entries[_wallet];
        require(_guardian != address(0) && _guardian != SENTINEL_GUARDIANS, "GS: invalid guardian");
        require(entry.guardians[_prevGuardian] == _guardian, "GS: invalid previous guardian");
        entry.guardians[_prevGuardian] = entry.guardians[_guardian];
        entry.guardians[_guardian] = address(0);
        entry.count--;
    }

    /**
     * @dev Allows to update the number of required confirmations by guardians.
     * @param _wallet The target wallet.
     * @param _threshold New threshold.
     */
    function changeThreshold(address _wallet, uint256 _threshold) external onlyModule(_wallet) {
        GuardianStorageEntry storage entry = entries[_wallet];
        // Validate that threshold is smaller than or equal to number of guardians.
        require(_threshold <= entry.count, "GS: threshold must be lower or equal to guardians count");
        if (entry.count == 0){
            require(_threshold == 0, "GS: threshold must be 0");
        }else{
            require(_threshold > 0, "GS: threshold cannot be 0");
        }
        entry.threshold = _threshold;
    }

    /**
     * @dev Checks if an account is a guardian for a wallet.
     * @param _wallet The target wallet.
     * @param _guardian The account.
     * @return true if the account is a guardian for a wallet.
     */
    function isGuardian(address _wallet, address _guardian) public view returns (bool) {
        return _guardian != SENTINEL_GUARDIANS && entries[_wallet].guardians[_guardian] != address(0);
    }

    /**
     * @dev Returns the number of guardians for a wallet.
     * @param _wallet The target wallet.
     * @return the number of guardians.
     */
    function guardiansCount(address _wallet) public view returns (uint256) {
        return entries[_wallet].count;
    }

    /**
     * @dev Retrieves the wallet threshold count.
     * @param _wallet The target wallet.
     * @return uint256 Threshold count.
     */
    function threshold(address _wallet) public view returns (uint256) {
        return entries[_wallet].threshold;
    }

    /**
     * @dev Gets the list of guaridans for a wallet.
     * @param _wallet The target wallet.
     * @return address[] list of guardians.
     */
    function getGuardians(address _wallet) public view returns (address[] memory) {
        GuardianStorageEntry storage entry = entries[_wallet];
        if (entry.count == 0){
            return new address[](0);
        }
        address[] memory array = new address[](entry.count);

        uint256 index = 0;
        address currentGuardian = entry.guardians[SENTINEL_GUARDIANS];
        while (currentGuardian != SENTINEL_GUARDIANS) {
            array[index] = currentGuardian;
            currentGuardian = entry.guardians[currentGuardian];
            index++;
        }
        return array;
    }

}

### Does this work on other EVMs? (If yes, please list at least 1 of them)

The same code deploys in zksync era testnet if I use the hardhat plugin, but doesn't deploy if I try to hash the bytecode according to zksync format and use the deployercontracts's create method to deploy

->"error":{"code":3,"message":"execution reverted: The code hash is not known"

### Description of What Your Contract Does

Just a storage contract

**Comment by elhajin** [comment]
hey,i think that's cause to deploy an account to zkEra , you have firs to mark the bytecode as known to able to deploy this code. in hardhat plugin this get done under the hood i think . to mark the code as known you sould call **KnownCodesStorage** [system contract](https://github.com/code-423n4/2023-03-zksync/tree/main/contracts), however  u can't call it  directly to mark the bytecode as known , since it's only can be called by the [byteCodeCompressor](https://github.com/code-423n4/2023-03-zksync/blob/main/contracts/BytecodeCompressor.sol)  [system contract](https://github.com/code-423n4/2023-03-zksync/tree/main/contracts),  
so , in summary you have to mark the bytecode as known first . then you call the **deployerContract** to deploy it  .

**Reply by 4NNNN** [reply]
I did take a look on it, the bytecodeCompressor is only callable by a bootloader address so there's no other way ?
function publishCompressedBytecode(
        bytes calldata _bytecode,
        bytes calldata _rawCompressedData
    ) external payable onlyBootloader returns (bytes32 bytecodeHash) {..}
    
Matterlabs zksync contract nodemodules has a different version !! But even though how do you encode the _rawCompressedData according to the format ?

**Reply by elhajin** [reply]
i think there is  a way , by sending an L1 transaction and the code will be marked as known , not sure how exactly. about the algorithm to compress bytecode (_rawCompressedData) i don't know it.. how they do it is by you send the bytecode in the factoryDeps field along with your tx, and  the operator (node) who have to do that , in the **publishCompressedBytecode** function the compressed bytecode (_rawCompressedData) is just get virified that the operator compressed it correctly and if so ,it get marked as known code.

**Reply by 4NNNN** [reply]
Yeah I agree, sending a l1 transaction inorder to deploy a contract in L2 sounds inefficient and gas costly !!

**Reply by 4NNNN** [reply]
@elhajin Hey elhajin, I have another doubt is there any alternative to handle ops funtionality in entrypoint contract to the bootloader in zksync era. I know the bootloader is ephemeral but is there any direct alternative to "handleops" functionality in zksync era ?

**Reply by elhajin** [reply]
sorry , don't get your question ,, if you  mean like in erc4337 , zkSync AA is  slightly different but you can think of the bootloader is the entrypoint

**Comment by ly0va** [comment]
Hi @4NNNN! Thank you for your question.

> doesn't deploy if I try to hash the bytecode 

Can you provide more details on this? If you only hash the bytecode but do not provide the bytecode preimage with it, then zkSync node won't be able to execute your code as it doesn't know what to execute (it cannot compute the preimage only having the hash, obv.). That's why you also have to put the whole bytecode as a factory dependency (`"factoryDep"`).

**Reply by 4NNNN** [reply]
I'm calling the createcall function in the deployer contract to deploy contracts directly from the frontend as the hardhat plugin runs only in hardhatruntime env and how do you want me to parse a preimage of the bytecode then hash it,  as well as use it as a calldata for the createcall params. Please provide me with a sample implementation !!

**Reply by 4NNNN** [reply]
@ly0va  I have another doubt is there any alternative to handle ops funtionality in entrypoint contract to the bootloader in zksync era. I know the bootloader is ephemeral but is there any direct alternative to "handleops" functionality in zksync era ?

**Reply by ly0va** [reply, marked as answer]
If you want to deploy a contract from the frontend, simply use the `ContractFactory` from our [zksync-web3 SDK](https://www.npmjs.com/package/zksync-web3). It has the same interface as `ContractFactory` in ethers.js.

There is no need to call the `ContractDeployer` system contract manually.

**Comment by varnerrobertv** [comment]
It seems like the contract deployment is failing because the code hash generated using the zkSync format is not recognized. This could be due to discrepancies in how the bytecode is hashed or formatted compared to zkSync's expectations. Double-check the bytecode hashing process and ensure it aligns with zkSync's requirements for contract deployment. Additionally, verify that the contract bytecode and its corresponding hash are correctly formatted and passed to the deployment function.

**Reply by 4NNNN** [reply]
Figured it out long time ago brother, thanks for your help though!

**Comment by fitzgibbonsdustin** [comment]
The error message "The code hash is not known" indicates that the deployed bytecode hash is not recognized by the zkSync system. This typically occurs when the bytecode hash generated during deployment does not match the expected format or when there are discrepancies in the deployment process.

To resolve this issue:
1. Ensure that you are correctly generating the bytecode hash according to the zkSync format. Double-check your implementation of the `generateBytecodeHash` function to ensure it produces the expected hash format.

2. Verify that the bytecode you are deploying is valid and matches the source code of your contract. Any discrepancies or modifications to the bytecode can cause mismatches in the generated hash.

3. Check for any errors or inconsistencies in the deployment process. Ensure that you are deploying the contract to the correct network and using the appropriate deployment parameters.

4. If you are using a deployment script or tool provided by zkSync, follow the recommended deployment procedures and ensure that all dependencies are properly configured.

By addressing these points and ensuring consistency in the deployment process, you should be able to resolve the issue and successfully deploy your contract on zkSync.

**Comment by phicurby** [comment]
The contract deploys successfully on zkSync Era Testnet using the Hardhat plugin but encounters an error when attempting to hash the bytecode according to zkSync format and deploying using the create method from the deployercontracts. The error message indicates "The code hash is not known." The contract itself is a storage contract.
